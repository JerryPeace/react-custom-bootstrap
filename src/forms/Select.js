import PropTypes from 'prop-types';
import React from 'react';
import {Basic} from './Basic';
import Label from './Label';
import LayoutHelper from './LayoutHelper';

class Select extends Basic {
    static propTypes = {
        labelText: PropTypes.node,
        getValue: PropTypes.func,
        onChange: PropTypes.func,
        items: PropTypes.array
    };

    static defaultProps = {
        bsClass: 'select',
        labelAlign: 'left',
        items: [],
        standalone: true,
        labelClassName: 'col-xs-2',
        wrapperClassName: 'col-xs-10'
    };

    constructor(props) {
        super(props);
        this.originalVal = ::this._initOriginalVal();
    }

    _initOriginalVal() {
        let value = '';
        if (typeof this.props.value !== 'undefined') {
            value = this.props.value;
        } else if (this.props.defaultValue) {
            value = this.props.defaultValue;
        } else if (this.props.items && this.props.items.length){
            value = this.props.items[0].value
        }
        return value;
    }

    getValue() {
        return this._inputEl.value;
    }

    renderSelect() {
        let {
            items,
            labelText,
            bsClass,
            labelAlign,
            standalone,
            labelClassName,
            wrapperClassName,
            value,
            defaultValue,
            ...rest
        } = this.props;

        let firstValue = defaultValue;
        if (!value && !defaultValue) {
            firstValue = this.originalVal;
        }

        return (
            <select
                ref={(el) => this._inputEl = el}
                key={'select'}
                value={value}
                defaultValue={firstValue}
                {...rest} >
                {items.map(function(opt, idx) {
                    const {text, ...rest} = opt;
                    return <option key={idx} {...rest}>{opt.text}</option>;
                })}
            </select>
        );
    }

    render() {
        return (
            <LayoutHelper FieldBase={this.renderSelect()} {...this.props} />
        );
    }
}

export default Select;
