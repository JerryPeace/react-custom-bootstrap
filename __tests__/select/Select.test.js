import React from 'react';
import { mount } from 'enzyme';
import toJSON from'enzyme-to-json';
import Select from 'forms/Select';

describe('Select', () => {
    let select;

    describe('with default settiings', () => {
        beforeEach(() => {
            select = mount(
                <Select
                    items={[
                        {value: 'opt1', text: 'This is Option1'},
                        {value: 'opt2', text: 'This is Option2'},
                        {value: 'opt3', text: 'This is Option3'}
                    ]}
                    labelText={'Select:'}
                    defaultValue={'opt2'} />
            );
        });

        test('should render corrently', () => {
            expect(toJSON(select)).toMatchSnapshot();
        });

        test('can get value', () => {
            expect(select.instance().getValue()).toBe('opt2');
        });

        test('can set value by props', () => {
            select.setProps({ value: 'opt1' });
            select.update();
            expect(select.instance().getValue()).toBe('opt1');
        });

    });

    describe('with empty items', () => {
        beforeEach(() => {
            select = mount(
                <Select
                    items={[]}
                    labelText={'Select:'} />
            );
        });

        test('should render corrently', () => {
            expect(toJSON(select)).toMatchSnapshot();
        });
    });

});
