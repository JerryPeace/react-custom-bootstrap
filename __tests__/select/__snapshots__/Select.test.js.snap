// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select with default settiings should render corrently 1`] = `
<Select
  bsClass="select"
  defaultValue="opt2"
  items={
    Array [
      Object {
        "text": "This is Option1",
        "value": "opt1",
      },
      Object {
        "text": "This is Option2",
        "value": "opt2",
      },
      Object {
        "text": "This is Option3",
        "value": "opt3",
      },
    ]
  }
  labelAlign="left"
  labelClassName="col-xs-2"
  labelText="Select:"
  standalone={true}
  wrapperClassName="col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <select
        defaultValue="opt2"
        value={undefined}
      >
        <option
          value="opt1"
        >
          This is Option1
        </option>
        <option
          value="opt2"
        >
          This is Option2
        </option>
        <option
          value="opt3"
        >
          This is Option3
        </option>
      </select>
    }
    bsClass="select"
    defaultValue="opt2"
    items={
      Array [
        Object {
          "text": "This is Option1",
          "value": "opt1",
        },
        Object {
          "text": "This is Option2",
          "value": "opt2",
        },
        Object {
          "text": "This is Option3",
          "value": "opt3",
        },
      ]
    }
    labelAlign="left"
    labelClassName="col-xs-2"
    labelText="Select:"
    singleCheckboxRadio={false}
    standalone={true}
    wrapperClassName="col-xs-10"
  >
    <div
      className=""
    >
      <Label
        FieldBase={
          <select
            defaultValue="opt2"
            value={undefined}
          >
            <option
              value="opt1"
            >
              This is Option1
            </option>
            <option
              value="opt2"
            >
              This is Option2
            </option>
            <option
              value="opt3"
            >
              This is Option3
            </option>
          </select>
        }
        className="col-xs-2"
        defaultValue="opt2"
        inline={false}
        items={
          Array [
            Object {
              "text": "This is Option1",
              "value": "opt1",
            },
            Object {
              "text": "This is Option2",
              "value": "opt2",
            },
            Object {
              "text": "This is Option3",
              "value": "opt3",
            },
          ]
        }
        labelAlign="left"
        text="Select:"
        wrapperClassName="col-xs-10"
      >
        <label
          className="col-xs-2"
        >
          Select:
        </label>
      </Label>
      <select
        className="form-control"
        defaultValue="opt2"
      >
        <option
          value="opt1"
        >
          This is Option1
        </option>
        <option
          value="opt2"
        >
          This is Option2
        </option>
        <option
          value="opt3"
        >
          This is Option3
        </option>
      </select>
    </div>
  </LayoutHelper>
</Select>
`;

exports[`Select with empty items should render corrently 1`] = `
<Select
  bsClass="select"
  items={Array []}
  labelAlign="left"
  labelClassName="col-xs-2"
  labelText="Select:"
  standalone={true}
  wrapperClassName="col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <select
        defaultValue=""
        value={undefined}
      >
        
      </select>
    }
    bsClass="select"
    items={Array []}
    labelAlign="left"
    labelClassName="col-xs-2"
    labelText="Select:"
    singleCheckboxRadio={false}
    standalone={true}
    wrapperClassName="col-xs-10"
  >
    <div
      className=""
    >
      <Label
        FieldBase={
          <select
            defaultValue=""
            value={undefined}
          >
            
          </select>
        }
        className="col-xs-2"
        inline={false}
        items={Array []}
        labelAlign="left"
        text="Select:"
        wrapperClassName="col-xs-10"
      >
        <label
          className="col-xs-2"
        >
          Select:
        </label>
      </Label>
      <select
        className="form-control"
        defaultValue=""
      />
    </div>
  </LayoutHelper>
</Select>
`;
